source/MRMesh/MRPointsComponents.cpp

```diff
@@ -0,0 +1,157 @@
+#include "MRPointsComponents.h"
+#include "MRPointCloud.h"
+#include "MRBitSet.h"
+#include "MRTimer.h"
+#include "MRPointsInBall.h"
+#include "MRProgressCallback.h"
+#include "MRPch/MRTBB.h"
+
+namespace MR
+{
+
+namespace PointCloudComponents
+{
+
+Expected<MR::VertBitSet> getLargestComponentsUnion( const PointCloud& pointCloud, float maxDist, int minSize, ProgressCallback pc /*= {}*/ )
+{
+    MR_TIMER
+
+    assert( maxDist > 0.f );
+    assert( minSize > 0 );
+    const auto& validPoints = pointCloud.validPoints;
+    ProgressCallback subPc = pc ? subprogress( pc, 0.f, 0.9f ) : pc;
+    auto unionStructsRes = getUnionFindStructureVerts( pointCloud, maxDist, nullptr, subPc );
+    if ( !unionStructsRes.has_value() )
+        return unexpectedOperationCanceled();
+    auto& unionStructs = *unionStructsRes;
+    auto allRoots = unionStructs.roots();
+
+    subPc = pc ? subprogress( pc, 0.9f, 0.95f ) : pc;
+    int counter = 0;
+    const float counterMax = float( validPoints.count() );
+    const int counterDivider = int( validPoints.count() ) / 100;
+    HashMap<VertId, int> root2size;
+    for ( auto v : validPoints )
+    {
+        ++root2size[allRoots[v]];
+        if ( !reportProgress( subPc, counter / counterMax, counter, counterDivider ) )
+            return unexpectedOperationCanceled();
+    }
+
+    subPc = pc ? subprogress( pc, 0.95f, 1.f ) : pc;
+    counter = 0;
+    VertBitSet result( validPoints.find_last() + 1 );
+    for ( auto v : validPoints )
+    {
+        if ( root2size[allRoots[v]] > minSize )
+            result.set( v );
+        if ( !reportProgress( subPc, counter / counterMax, counter, counterDivider ) )
+            return unexpectedOperationCanceled();
+    }
+
+    return result;
+}
+
+Expected<UnionFind<MR::VertId>> getUnionFindStructureVerts( const PointCloud& pointCloud, float maxDist, const VertBitSet* region /*= nullptr*/, ProgressCallback pc /*= {}*/ )
+{
+    MR_TIMER
+
+    VertBitSet vertsRegion = pointCloud.validPoints;
+    if ( region )
+        vertsRegion &= *region;
```

Comment:
```suggestion
    const VertBitSet & vertsRegion = region ? *region : pointCloud.validPoints;
```

Reply:
It is a user error then. We do not want to make slower algorithms (or copy more) because of errors.